File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* helper(vector<int>& preorder,int l,int r,vector<int>& inorder,unordered_map<int,int>& m,int& idx){
        if(l>r){
            return NULL;
        }
        int val=preorder[idx++];
        TreeNode* root=new TreeNode(val);
        if(l==r){
            return root;
        }
        int inRoot=m[val];
        root->left=helper(preorder,l,inRoot-1,inorder,m,idx);
        root->right=helper(preorder,inRoot+1,r,inorder,m,idx);
        return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int,int>m;
        int sz=inorder.size();
        for(int i=0;i<sz;i++){
            m[inorder[i]]=i;
        }
        int idx=0;
        TreeNode* root=helper(preorder,0,sz-1,inorder,m,idx);
        return root;
    }
};

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int idx=0;
    Map<Integer,Integer> m=new HashMap<>();

    public TreeNode build(int[] preorder,int l,int r){
        if(l>r){
            return null;
        }

        int root=preorder[idx++];
        TreeNode rootV=new TreeNode(root);

        int rootidx=m.get(root);

        rootV.left=build(preorder,l,rootidx-1);
        rootV.right=build(preorder,rootidx+1,r);

        return rootV;

    }

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        
        for(int i=0;i<inorder.length;i++){
            m.put(inorder[i],i);
        }

        return build(preorder,0,inorder.length-1);
    }
}